---
title: "Obtaining cellular network maps with SingleCellSignalR"
date: "`r format(Sys.Date(), '%m/%d/%Y')`"
author:
- name: "Simon Cabello-Aguilar"
  affiliation: "Institut de Recherche en Cancérologie de Montpellier, Inserm, Montpellier, France ; Institut régional du Cancer Montpellier, Montpellier, France ; Université de Montpellier, Montpellier, France"
- name: "Jacques Colinge"
  affiliation: "Institut de Recherche en Cancérologie de Montpellier, Inserm, Montpellier, France ; Institut régional du Cancer Montpellier, Montpellier, France ; Université de Montpellier, Montpellier, France"
output: 
  rmarkdown::html_document:
    highlight: pygments
    toc: true
vignette: >
  %\VignetteIndexEntry{my-vignette}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{=html}
<style type="text/css">
.smaller {
  font-size: 10px
}
</style>
```

```{r setup, echo=FALSE, results="hide"}
knitr::opts_chunk$set(tidy = FALSE,
                      cache = FALSE,
                      dev = "png",
                      message = FALSE, error = FALSE, warning = TRUE)
```	

# Introduction

## What is `SingleCellSignalR`?

This guide provides an overview of the SingleCellSignalR package, a comprehensive framework to obtain cellular network maps from scRNA-seq data. SingleCellSignalR comes with a complete pipeline integrating existing methods to cluster individual cell transcriptomes and identify cell subpopulations as well as novel cellular network-specific algorithms. More advanced users can substitute their own logic or alternative tools at various stages of data processing. SingleCellSignalR also maps cell subpopulation internal network linked to genes of interest through the integration of regulated KEGG and Reactome pathways together with ligands and receptors involved in inferred cell-cell interactions. The cellular networks can be exported in text files and graphML objects to be further explored with Cytoscape (www.cytoscape.org), yEd (www.yworks.com), or similar software tools.  

## Starting point

Independently of the chosen scRNA-seq platform, deep or shallower, data comes as a table of read or unique molecule identifier (UMI) counts, one column per individual cell and one row per gene. Initial processing is required to prepare such data for subsequent analysis and we decided to propose a generic solution for the sake of convenience, though users can easily substitute their own computations. Gene names (HUGO symbols) are provided in the first column of the table. 

Each analysis is organized around a working directory (or project folder):  

![][id0]

[id0]: ../man/figures/directory.png

# Quick Start

## Data pre-processing workflow

Here we will describe the use of the `SingleCellSignalR` package from a raw read counts matrix to the obtaining of an object suitable for analyzing ligand-receptor (LR) interactions and their visualization. 

Where do the file used for example come from?
  
### Data loading

There are two ways of loading the data in `SingleCellSignalR` through the `dataPrepare()` function:
- the read counts matrix is opened in the R environment: the matrix is given as the `file` parameter.
- the file containing the read counts is in the working directory: the file's name is given as the `file` parameter. 


Data processing can then start. By default the `dataPrepare()` function eliminates non expressed genes before performing read count normalization through a method by quantile. 
When not working on humans, the species needs to be indicated through the `species` parameter and an ortholog genes correspondence will be performed.

```{r, echo=TRUE, eval=TRUE}
library(SingleCellSignalR)
example_dataset_path <- system.file("extdata", "scRNAseq_dataset.txt", package = "SingleCellSignalR")
print(example_dataset_path)
example_dataset <-  read.csv(example_dataset_path, sep = "\t")
print(example_dataset)
# data <- as.data.frame(data)
data <- example_dataset
```

```{r, echo=TRUE, eval=TRUE}
object <- dataPrepare(file = data)
```

This first function creates an object that stores the data used for ligand-receptor analysis. It contains the read counts matrix, the species and eventually information upon orthologs genes.

### Clustering

Two options are possible for clustering:

* **statistically based approach** through *kmeans* or *simlr* algorithm to cluster cells based on gene expression. This is done through the `cellClustering()` function.

* **cell based approach** using a marker table to classify each cell to a specific celltype. This is done through the `cellClassifying()` function

#### `cellClustering()`

Normalized data are submitted to a clustering algorithm to identify cell subpopulations:  

```{r, echo=TRUE, eval=FALSE}
object <- cellClustering(obj = object, n = 10, method = "simlr")
```

We set the method argument to `simlr`, which caused the **SIMLR()** function of the SIMLR package [1] to be used. The **SIMLR_Estimate_Number_of_Clusters()** function determined the number of clusters, between 2 and n (n=10 above). 
Alternatively the `kmeans` method can be used for clustering.

At this step, specific annotation can be attributed to each cluster by specifying a `classification` method:
* **by cluster**: mean gene expression for each cluster is compared with the signatures described in the `markers` table. The `markers()` function enables the user to select specific celltype in the default markers table. A personal markers table can also be inputted.
* **singler**: the package `SingleR` is used with the reference dataset from `HumanPrimaryCellAtlasData`.

A 2D tsne projection can be calculated and plotted (default `plot = TRUE`).

```{r, echo=TRUE, eval=FALSE}
markers = markers(c("immune","tme"))
object <- cellClustering(obj = object, n = 10, method = "simlr", 
                classification = "by cluster",
                         markers = markers)
```

```{r, echo=TRUE, eval=FALSE}
object <- cellClustering(obj = object, n = 10, method = "simlr", classification = "singler")
```

Cluster annotations are stored in the object's `cluster` slot, alongside the method used, and the eventual markers table.

#### `cellClassifying()`

The gene expression for each cell is compared with the signatures described in the `markers` table. If no projection method is available in the object (userly inputted or `cellClustering()` run previously), tsne coordinates will be calculated at this step.

```{r, eval=FALSE}
markers <- markers(c("immune","tme"))
object.b <- cellClassifying(obj = object, markers = markers)
```

Cluster annotations are stored in the object's `cluster` slot, alongside the method used, and the eventual markers table.

## Using our own processed data

The pre-processing workflow presented previously is optionnal and can be skipped when working on already normalized and clustered data.

The object needs to be created with the `dataPrepare()` function. To avoid normalization, set the `normalize` argument to "none".

```{r, echo=TRUE, eval = FALSE}
object <- dataPrepare(file = file.normalized, normalize = "none")
```

If clustering and cell projection is already available it must be added to the object with the `addClustering()` function. If no cell projection coordinates are given, no plots will be outputted.
When working with Seurat object, the clustering can be accessed via `seurat.object@metadata$seurat_clusters` and the cell projection via `seurat.object@reductions$tsne@cell.embeddings` (*tsne* or any other projection method used, e.g. *umap*).

```{r, echo=TRUE, eval = FALSE}
object <- addClustering(obj = object, cluster.id = clusters)
```

## Differential expression

Next, differentially expressed genes in one cluster compared to the others are identified using the `dgeCluster()` function.  which relies on **edgeR**. A result table is automatically created in the *cluster-analysis* folder. DGE data and parameters are stored in the **SCSRDataModel** object's `dge.cluster` slot.

```{r, eval=FALSE}
object <- dgeCluster(obj = object)
```

## Ligand-receptor analysis

Once the preliminary steps illustrated above are completed, **SingleCellSignalR** can be used to generate cellular interaction lists using the `cellSignaling()` function:  

```{r, eval=FALSE}
object.int <- cellSignaling(obj = object)
```

### Visualizations 

A summary of the interactions between cell clusters can be output in the form of a chord diagram by the  `visualize_interactions()` function:  

```{r, eval=FALSE, fig.height=7, fig.width=7}
visualizeInteractions(obj = object.int, dm = object)
```

This function will create a *plot* in the R plot window.  

The details of the interactions between two clusters, for example cluster 1 and 2, can also be shown in the plot window with the `visualize_interactions()` function. Note that in the example below we ask for the display of two pairs of cell clusters, pair 1 that contains interactions from cluster 1 to 2, and pair 4 from cluster 2 to 1. (`names(LRinter(object.inter))` returns the cell cluster names in each pair, see function `visualize_interactions()` details.)   

```{r, eval=FALSE, fig.height=7, fig.width=7}
visualizeInteractions(obj = object.int, dm = object,show.in=c(1,4))
```   

*Remark: We observe that in the chord diagrams above, the “specific” interactions were highlighted with a thick black line.*

And these plots can be saved into pdf files in the *images* folder using the `write.in` argument of the `visualize_interactions()` function.  
```r , eval=FALSE
visualize_interactions(obj = object.int, dm = object,write.in=c(1,4))
```

The ligand-receptor scores within clusters can be represented through the function `interactionDotPlot()`:

```{r, eval=FALSE, fig.height=7, fig.width=7}
interactionDotPlot(obj = object.int, interaction = 4)
``` 

Many other visualization functions are available to plot interaction results.

### Visualization of pathway(s) of interest

The different visualization functions contain an argument `pathways` that can be set to specify one or several pathways of interest. The pathways need to be contained in the *ReactomeKEGG* database. The function `findpathways()` can be used to access all pathways' names containing a specific keyword.

```{r, echo=TRUE, eval = FALSE}
pathways <- findPathways("interleukin")
```

The specific pathway(s) name(s) can then be used in the different functions: `visualizeInteractions()`, `interNetwork()`, `intraNetwork()` and `interactionDotPlot()`.

## Adding own LR interaction to database

It is possible to add specific LR interactions to be studied. When using the `cellSignaling()` function, specify the LR interactions as a dataframe through the `addLR` argument:

```{r, echo=TRUE, eval = FALSE}
LRadded <- dataframe(ligand = paste("Ligand ",seq_len(10)), receptor = paste("Receptor ",seq_len(10)))
object.int <- cellSignaling(obj = object, addLR = LRadded)
```

## Switching the LR database 

It is possible to switch the ligand-receptor interaction database from LRdb to any other. When using the `cellSignaling()` function, specify the LR database as a dataframe (with 2 columns *ligand** and receptor) through the `switchDB` argument:

```{r, echo=TRUE, eval = FALSE}
databaseLR_new <- dataframe(ligand = paste("Ligand ",seq_len(10)), receptor = paste("Receptor ",seq_len(10)))
object.int <- cellSignaling(obj = object, switchDB = databaseLR_new)
```

<span style="color:white">red</span>   
<span style="color:white">red</span>   
<span style="color:white">red</span> 

------

# Examples of use 

**SingleCellSignalR** package functions have many arguments parameters that can be changed by the user to fit her needs (see Reference Manual for more details). Furthermore, several handy functions that were not illustrated above are provided to generate additional plots or reports.  

## Marker analysis on a cancer dataset

For this example we use the scRNAseq dataset from Tirosh et al. [4]. We use only the data from patient 80.

```{r, eval=TRUE}
data(Patient80, package = "SingleCellSignalR")

object <- dataPrepare(file = data, lower=0.05,upper=0.0155)
```

*Remark: One can notice that the zero rate is lower than in the previous example which reflects the fact that the sequencing is deeper.*  

We know that this dataset is composed of melanoma cells and their microenvironment, we hence define our markers table using the `markers()` function.  

```{r, eval=FALSE}
my.markers <- markers(category = c("immune", "tme", "melanoma"))
head(my.markers)
```

Let us perform the clustering. For this example, we set the `method` argument to *"kmeans"* and the *n* argument to 12. We use the markers table created above to assign a celltype to each cluster.

```{r, eval=FALSE}
object <- cellClustering(obj = object, n = 12, method = "simlr", markers = my.markers, classification = "by cluster")
```

We can compare the results with the one obtained through cell annotation:
```{r, eval=FALSE}
object.b <- cellClassifying(obj = object, markers = my.markers)
```

Differential gene expression can be performed to find specific genes expressed in each cluster (than can then be used to define specific ligand-receptor interactions):

```{r, eval=FALSE}
object <- dgeCluster(obj = object)
```

The most variable interactions between cluster can be analysed and plotted through the `mv_interactions()` function: 
```{r, eval=FALSE, fig.height=7, fig.width=7}
mvInteractions(obj = object)
```  

It is possible to look at specific genes' coupled expression using the `expression_plot_2()` function.
```{r, eval=FALSE}
expressionPlot2(obj = object, "AREG", "EGFR")
```  

With the function `expression_plot()` it is possible to look at the expression of one gene within each cell.
```{r, eval=FALSE}
expressionPlot(obj = object, "ANXA1")
```  

LR interactions are calculated between the different clusters (interaction type is by default `paracrine`):
```{r, eval=FALSE}
object.int <- cellSignaling(obj = object)
```

And then visualized, eventually specifying pathway(s) to study:

```{r, eval=FALSE}
pathways <- findPathways("downregulation")
p <- pathways[2,1]
visualizeInteractions(obj = object.int, dm = object, show.in = c(3,1), pathway = p)
```

```{r, eval=FALSE}
interactionDotPlot(obj = object.int, dm = object, interaction = 3, pathway = p)
```

An intercellular network can also be generated to map the ligand/receptor interactions linked with specific pathway(s) calling the `inter_network()` function:  

```{r, eval=FALSE}
inter.net <- interNetwork(obj = object.int, dm = object, pathway = p)
```

We can be interested in genes participating in pathways with a receptor of interest inside a cluster of interest. Let us say *EGFR* in the second cluster.

```{r, eval=FALSE}
g <- "EGFR"
c <- cluster(object)$names[2]
intra.net <- intraNetwork(dm = object, obj = object.int, goi = g, coi = c, pathway = p)
```

At this point the intercellular network have been generated and exported in text and graphML formats in the *networks* folder.   

## Mouse dataset

For this example we use the scRNAseq dataset from `recount3`'s *SRP066963* project. Because the dataset is not a human dataset the species studied must be given upon preparing the data: 

```{r}
matrix_path <- system.file("extdata", "mouse_scRNAseq_dataset.txt", package = "SingleCellSignalR")
data <-  read.csv(matrix_path, sep = "\t")
print(data)

object <- dataPrepare(file = data, lower=0.05,upper=0.0155, species = "mmusculus")
```

Mouse genes are converted to human genes (initial mouse genes are found in `ncounts(object)$initial.orthologs`). Genes in the dataset are compared with the ligand receptor database (**LRdb**) to verify their presence.

We apply clustering to the data and use the `singler` package to annotate them:

```{r, eval=FALSE}
object <- cellClustering(obj = object, method = "simlr", classification = "singler")
```

Differential gene expression can be performed to find specific genes expressed in each cluster (than can then be used to define specific ligand-receptor interactions):

```{r, eval=FALSE}
object <- dgeCluster(obj = object)
```

The most variable interactions between cluster can be analysed and plotted through the `mv_interactions()` function: 
```{r, eval=FALSE, fig.height=7, fig.width=7}
mvInteractions(obj = object)
```  

LR interactions are calculated between the different clusters (interaction type is by default `paracrine`):
```{r, eval=FALSE}
object.int <- cellSignaling(obj = object)
```

<span style="color:white">red</span>   
<span style="color:white">red</span>   
<span style="color:white">red</span> 

Thank you for reading this guide and for using **SingleCellSignalR**.

----

# Session information

```{r session-info}
sessionInfo()
```

# *References*

1. Wang B, Zhu J, Pierson E, Ramazzotti D, Batzoglou S. Visualization and analysis of single-cell RNA-seq data by kernel-based similarity learning. Nat Methods. 2017;14:414-6.

2. McCarthy DJ, Chen Y, Smyth GK. Differential expression analysis of multifactor RNA-Seq experiments with respect to biological variation. Nucleic Acids Res. 2012;40:4288-97.

3. 8k PBMCs from a Healthy Donor [Internet]. 2017. Available from: https://support.10xgenomics.com/single-cell-gene-expression/datasets/2.1.0/pbmc8k

4. Tirosh I, Izar B, Prakadan SM, Wadsworth MH, Treacy D, Trombetta JJ, et al. Dissecting the multicellular ecosystem of metastatic melanoma by single-cell RNA-seq. Science. 2016;352:189-96.
 
 
















